File: 653.cpp
/*
Given a Binary Search Tree and a target number, return true if there exist two elements in the BST such that their sum is equal to the given target.

Example 1:
Input: 
    5
   / \
  3   6
 / \   \
2   4   7

Target = 9

Output: True
Example 2:
Input: 
    5
   / \
  3   6
 / \   \
2   4   7

Target = 28

Output: False
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    
    TreeNode* next(stack<TreeNode*>& s, TreeNode* &cur, bool inc) {
        while (cur) {
            s.push(cur);
            cur = inc ? cur->left : cur->right;
        }
        
        TreeNode* npt = s.top();
        s.pop();
        
        cur = inc ? npt->right : npt->left;
        return npt;
    }
    
    bool findTarget(TreeNode* root, int k) {
        if (!root) return false;
        
        stack<TreeNode*> ls, rs;
        TreeNode *icur = root, *dcur = root;
        TreeNode *p = next(ls, icur, true), *q = next(rs, dcur, false);
        
        while (p->val != q->val) {
            if (p->val + q->val == k) return true;
            else if (p->val + q->val < k) {
                // move p
                p = next(ls, icur, true);
            }
            else {
                // move q
                q = next(rs, dcur, false);
            }
        }
        return false;
    }
};

class Solution {
    public boolean findTarget(TreeNode root, int k) {
        Set<Integer> hashset = new HashSet<>();
        return find(root, k, hashset);
    }
    
    private boolean find(TreeNode root, int k, Set<Integer> hashset) {
        if (root == null) return false;
        if (hashset.contains(k - root.val)) return true;
        hashset.add(root.val);
        
        return find(root.left, k, hashset) || find(root.right, k, hashset);
    }
}

File: 653.java
/*
Given an integer array with no duplicates. A maximum tree building on this array is defined as follow:

The root is the maximum number in the array.
The left subtree is the maximum tree constructed from left part subarray divided by the maximum number.
The right subtree is the maximum tree constructed from right part subarray divided by the maximum number.
Construct the maximum tree by the given array and output the root node of this tree.

Example 1:
Input: [3,2,1,6,0,5]
Output: return the tree root node representing the following tree:

      6
    /   \
   3     5
    \    / 
     2  0   
       \
        1
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode constructMaximumBinaryTree(int[] nums) {
        if (nums == null || nums.length == 0) return null;
        return construct(nums, 0, nums.length-1);
    }
    
    private TreeNode construct(int[] nums, int left, int right) {
        if (left > right) return null;
        int maxIndex = left;
        for (int i = left; i <= right; i++)
            if (nums[i] > nums[maxIndex])
                maxIndex = i;
        TreeNode root = new TreeNode(nums[maxIndex]);
        root.left = construct(nums, left, maxIndex - 1);
        root.right = construct(nums, maxIndex+1, right);
        return root;
    }
}
